src/scanner.ts

```diff
@@ -12,35 +12,48 @@ export interface ScanResult {
   ruleName: string;
   description: string;
   total: number;
-  ids: string[];
+  resources: {id: string}[];
 }
 
 export class Scanner {
-  private _rules: Rule[] = [];
-
-  async scan(ruleType: Rule['type'], target: string) {
-    if (!this._rules.length) await this.loadRulesFromFile();
-    return this._executeRules(ruleType, target);
+  async scan(context: RuleContext, target: RuleContext['target']) {
+    switch (context.type) {
+      case 'resourceGraph': {
+        context.target = target as ResourceGraphRuleContext['target'];
+        return ResourceGraphRule.execute(context);
+      }
+      case 'dummy': {
+        context.target = target as DummyRuleContext['target'];
+        return DummyRule.execute(context);
+      }
+    }
   }
 
-  async loadRulesFromFile(filePath = '../../rules.json') {
-    const absPath = path.join(__dirname, filePath);
-    const data = await fsPromises.readFile(absPath, 'utf8');
-    this._rules = JSON.parse(data);
+  getContextByType(type: RuleContext['type'], contexts: RuleContext[]) {
+    return contexts.filter(r => r.type === type)[0];
   }
 
-  private _executeRules(type: Rule['type'], target: string) {
-    const filteredRules = this._rules.filter(r => r.type === type);
-    switch (type) {
-      case 'resourceGraph': {
-        return ResourceGraphRule.execute(
-          filteredRules as IResourceGraphRule[],
-          target
-        );
-      }
-      case 'dummy': {
-        return DummyRule.execute(filteredRules as IDummyRule[]);
+  filterRulesByName(names: string[], context: RuleContext) {
```

Comment:
We should make changes here as well. We've got more rule/context mutation going on, and the implementation isn't straightforward to read. I think what we are saying is "for all the rules that we have, only run the ones that are explicitly listed". So that's something like 

```javascript
const rulesToRun = rules.filter(r => names.includes(r));
```

and then there's the probably-an-error case where we say "run this rule" and it's not defined

```javascript
const names = ['one', 'ten'];
const rules = ['one', 'two', 'three'];

// 'ten' isn't listed, so we can't run all the rules that you want
// We probably fail the app here, otherwise you think everything's good when in reality we just didn't do what you asked
const rulesToRun = filterRules(rules, names);
```